### PR TITLE
feat(menu): Convert JS to TypeScript

### PR DESCRIPTION
**Pull request by acdvorak**
Refs #4225

### CODE REVIEW

File: packages/mdc-base/component.ts [hunk 1]
```diff
@@ -82,16 +85,16 @@ class MDCComponent<FoundationType extends MDCFoundation> {
    * Wrapper method to add an event listener to the component's root element. This is most useful when
    * listening for custom events.
    */
-  listen(evtType: string, handler: EventListener) {
-    this.root_.addEventListener(evtType, handler);
+  listen<T extends Event>(evtType: string, handler: EventListener | CustomEventListener<T>) {
```
Comment (mmalerba) [inline packages/mdc-base/component.ts]: We recently [added some types](https://github.com/material-components/material-components-web/blob/feat/typescript/packages/mdc-dom/events.ts) for dealing with standard browser events, if you also want to give users the ability to specify the event type in the case of custom events, you could use typescript's function overloading to support it while still filling in the right type for standard events [(example)](http://www.typescriptlang.org/play/#src=type%20EventType%20%3D%20keyof%20GlobalEventHandlersEventMap%3B%0D%0Atype%20SpecificEventListener%3CK%20extends%20EventType%3E%20%3D%20(evt%3A%20GlobalEventHandlersEventMap%5BK%5D)%20%3D%3E%20void%3B%0D%0Atype%20CustomEventListener%3CE%20extends%20Event%3E%20%3D%20(evt%3A%20E)%20%3D%3E%20void%3B%0D%0A%0D%0Afunction%20listen%3CK%20extends%20EventType%3E(event%3A%20K%2C%20listener%3A%20SpecificEventListener%3CK%3E)%3B%0D%0Afunction%20listen%3CE%20extends%20Event%3E(event%3A%20string%2C%20listener%3A%20CustomEventListener%3CE%3E)%3B%0D%0Afunction%20listen(event%2C%20listener)%20%7B%20%2F*%20implementation%20here%20*%2F%20%7D%0D%0A%0D%0Ainterface%20WhateverEvent%20extends%20Event%20%7B%0D%0A%20%20%20%20stuff%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Alisten('click'%2C%20e%20%3D%3E%20e.clientX)%3B%0D%0Alisten%3CWhateverEvent%3E('whatever'%2C%20e%20%3D%3E%20e.stuff)%3B)

Comment (acdvorak) [reply packages/mdc-base/component.ts]: Great suggestion, @mmalerba! PTAL 😀

File: packages/mdc-menu/foundation.ts [hunk 2]
```diff
@@ -21,52 +21,46 @@
  * THE SOFTWARE.
  */
 
+// TODO(acdvorak): Create PR to update all MDC imports to use {} syntax rather than default export
```
Comment (acdvorak) [inline packages/mdc-menu/foundation.ts]: Create PR for this

File: packages/mdc-menu/types.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * Event properties used by the adapter and foundation.
+ */
+export interface MenuItemEventDetail {
+  index: number;
+}
+
+/**
+ * Event properties specific to the default component implementation.
+ */
+export interface MenuItemComponentEventDetail extends MenuItemEventDetail {
```
Comment (acdvorak) [inline packages/mdc-menu/types.ts]: Maybe rename to `DefaultMenuItemEventDetail`?

File: packages/mdc-menu/README.md [hunk 4]
```diff
@@ -227,22 +227,22 @@ Method Signature | Description
...
 `removeAttributeFromElementAtIndex(index: number, attr: string) => void` | Removes the `attr` attribute from the element at the `index` specified.
-`elementContainsClass(element: HTMLElement, className: string) => boolean` | Returns true if the `element` contains the `className` class.
+`elementContainsClass(element: Element, className: string) => boolean` | Returns true if the `element` contains the `className` class.
 `closeSurface() => void` | Closes the menu surface.
-`getElementIndex(element: HTMLElement) => number` | Returns the `index` value of the `element`.
-`getParentElement(element: HTMLElement) => ?HTMLElement` | Returns the `.parentElement` element of the `element` provided.
-`getSelectedElementIndex(element: HTMLElement) => number` | Returns the `index` value of the element within the selection group provided, `element` that contains the `mdc-menu-item--selected` class.
+`getElementIndex(element: Element) => number` | Returns the `index` value of the `element`.
+`getParentElement(element: Element) => ?Element` | Returns the `.parentElement` element of the `element` provided.
```
Comment (moog16) [inline packages/mdc-menu/README.md]: is the `?` prefix a closure thing?

File: packages/mdc-menu/README.md [hunk 5]
```diff
@@ -227,22 +227,22 @@ Method Signature | Description
...
+`getSelectedElementIndex(element: Element) => number` | Returns the `index` value of the element within the selection group provided, `element` that contains the `mdc-menu-item--selected` class.
 `notifySelected(index: number) => void` | Emits a `MDCMenu:selected` event for the element at the `index` specified.
 
 ### `MDCMenuFoundation`
 
 Method Signature | Description
 --- | ---
 `handleKeydown(evt: Event) => void` | Event handler for the `keydown` events within the menu.
-`handleItemAction(listItem: !HTMLElement) => void` | Event handler for list's action event.
+`handleItemAction(listItem: !Element) => void` | Event handler for list's action event.
```
Comment (moog16) [inline packages/mdc-menu/README.md]: Same here - the `!` seems like closure syntax. I think we can remove.

File: packages/mdc-menu/README.md (hunk 4 above)
Comment (moog16) [inline packages/mdc-menu/README.md]: Also TS `lib.dom.d.ts` states it is an `HTMLElement`

```
/**
     * Returns the parent element.
     */
    readonly parentElement: HTMLElement | null;

```

File: packages/mdc-menu/adapter.ts [hunk 6]
```diff
@@ -21,95 +21,64 @@
...
+  getElementIndex(element: Element): number;
 
   /**
-   * Returns the parentElement of the provided element.
-   * @param {?HTMLElement} element
-   * @return {?HTMLElement} parentElement of the element provided.
+   * @return The parentElement of the provided element.
    */
-  getParentElement(element) {}
+  getParentElement(element: Element): Element | null;
```
Comment (moog16) [inline packages/mdc-menu/adapter.ts]: according to TS `Node` class it is an HTMLElement that is returned

File: packages/mdc-menu/index.ts [hunk 7]
```diff
@@ -22,57 +22,51 @@
...
 
-    /** @private {!Function} */
-    this.handleItemAction_;
+  private menuSurfaceFactory_!: MenuSurfaceFactory; // assigned in initialize()
+  private listFactory_!: ListFactory; // assigned in initialize()
 
-    /** @private {!Function} */
-    this.afterOpenedCallback_;
-  }
+  private menuSurface_!: MDCMenuSurface; // assigned in initialize()
```
Comment (moog16) [inline packages/mdc-menu/index.ts]: I don't think the `//` comments are necessary. 

But the `menuSurface_` and `list_` are assigned in `initialSyncWithDom`

File: packages/mdc-menu/index.ts [hunk 8]
```diff
@@ -151,50 +140,41 @@ class MDCMenu extends MDCComponent {
...
     const list = this.items;
     if (list.length > 0) {
-      list[0].focus();
+      (list[0] as HTMLElement).focus();
     }
   }
 
-  /** @return {!MDCMenuFoundation} */
+  /**
+   * @return {!MDCMenuFoundation}
```
Comment (moog16) [inline packages/mdc-menu/index.ts]: change to `@return MDCMenuFoundation`

File: test/unit/mdc-menu/mdc-menu.test.js [hunk 9]
```diff
@@ -26,10 +26,10 @@ import bel from 'bel';
 import td from 'testdouble';
 import domEvents from 'dom-events';
 
-import {MDCMenu, MDCMenuFoundation} from '../../../packages/mdc-menu/index';
-import {Corner} from '../../../packages/mdc-menu-surface/constants';
-import {MDCListFoundation} from '../../../packages/mdc-list/index';
-import {MDCMenuSurfaceFoundation} from '../../../packages/mdc-menu-surface/foundation';
+import {MDCMenu, MDCMenuFoundation} from '../../../packages/mdc-menu/index.ts';
```
Comment (moog16) [inline test/unit/mdc-menu/mdc-menu.test.js]: I don't think adding the `.ts` extension is necessary. It should be resolved by webpack.

File: test/unit/mdc-menu/menu.foundation.test.js [hunk 10]
```diff
@@ -26,10 +26,10 @@ import td from 'testdouble';
 import {verifyDefaultAdapter} from '../helpers/foundation';
 import {setupFoundationTest} from '../helpers/setup';
 import {install as installClock} from '../helpers/clock';
-import {MDCMenuFoundation} from '../../../packages/mdc-menu/foundation';
-import MDCListFoundation from '../../../packages/mdc-list/foundation';
-import {cssClasses, strings} from '../../../packages/mdc-menu/constants';
-import {numbers} from '../../../packages/mdc-menu-surface/constants';
+import {MDCMenuFoundation} from '../../../packages/mdc-menu/foundation.ts';
```
Comment (moog16) [inline test/unit/mdc-menu/menu.foundation.test.js]: here too

File: test/unit/mdc-select/mdc-select-enhanced.test.js [hunk 11]
```diff
@@ -32,8 +32,8 @@ import {MDCRipple, MDCRippleFoundation} from '../../../packages/mdc-ripple/index
 import {MDCSelect} from '../../../packages/mdc-select/index';
 import {cssClasses, strings} from '../../../packages/mdc-select/constants';
 import {MDCNotchedOutline} from '../../../packages/mdc-notched-outline/index';
-import {MDCMenu, MDCMenuFoundation} from '../../../packages/mdc-menu/index';
-import {MDCMenuSurfaceFoundation} from '../../../packages/mdc-menu-surface/index';
+import {MDCMenu, MDCMenuFoundation} from '../../../packages/mdc-menu/index.ts';
```
Comment (moog16) [inline test/unit/mdc-select/mdc-select-enhanced.test.js]: here too

File: packages/mdc-menu/index.ts (hunk 7 above)
Comment (kfranqueiro) [reply packages/mdc-menu/index.ts]: The comments are a good idea for anything using `!` to explain why it's there.

File: packages/mdc-menu/README.md (hunk 4 above)
Comment (acdvorak) [reply packages/mdc-menu/README.md]: Done

File: packages/mdc-menu/README.md (hunk 5 above)
Comment (acdvorak) [reply packages/mdc-menu/README.md]: Done

File: packages/mdc-menu/adapter.ts (hunk 6 above)
Comment (acdvorak) [reply packages/mdc-menu/adapter.ts]: `lib.dom.d.ts` is wrong 😛 

* https://dom.spec.whatwg.org/#parent-element
* https://developer.mozilla.org/en-US/docs/Web/API/Node/parentElement

File: packages/mdc-menu/README.md (hunk 4 above)
Comment (acdvorak) [reply packages/mdc-menu/README.md]: `lib.dom.d.ts` is wrong 😛 

* https://dom.spec.whatwg.org/#parent-element
* https://developer.mozilla.org/en-US/docs/Web/API/Node/parentElement

File: packages/mdc-menu/index.ts (hunk 7 above)
Comment (acdvorak) [reply packages/mdc-menu/index.ts]: Done

File: packages/mdc-menu/index.ts (hunk 8 above)
Comment (acdvorak) [reply packages/mdc-menu/index.ts]: Removed comment in favor of TypeScript syntax

File: test/unit/mdc-menu/mdc-menu.test.js (hunk 9 above)
Comment (acdvorak) [reply test/unit/mdc-menu/mdc-menu.test.js]: Removed

File: test/unit/mdc-menu/menu.foundation.test.js (hunk 10 above)
Comment (acdvorak) [reply test/unit/mdc-menu/menu.foundation.test.js]: Removed

File: test/unit/mdc-select/mdc-select-enhanced.test.js (hunk 11 above)
Comment (acdvorak) [reply test/unit/mdc-select/mdc-select-enhanced.test.js]: Removed

File: packages/mdc-menu/types.ts (hunk 3 above)
Comment (acdvorak) [reply packages/mdc-menu/types.ts]: Done

File: packages/mdc-menu/foundation.ts (hunk 2 above)
Comment (acdvorak) [reply packages/mdc-menu/foundation.ts]: Done

File: packages/mdc-menu/adapter.ts (hunk 6 above)
Comment (moog16) [reply packages/mdc-menu/adapter.ts]: ha - ya that makes sense. I would think they want it to be Element. Ok let's leave it for now.

File: packages/mdc-menu/types.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
+/**
+ * @license
+ * Copyright 2018 Google Inc.
```
Comment (moog16) [inline packages/mdc-menu/types.ts]: 2019

Comment (acdvorak) [reply packages/mdc-menu/types.ts]: Done

**Pull request merged**